File: 백준/Gold/3079. 입국심사/입국심사.cc
#include <iostream>
#include <algorithm>
#include <vector>

#define MAX_SIZE 10000000000

using namespace std;

long long n, m;

bool check_count(long long target, vector<long long> &vec) {
	long long cnt = 0;
	for (auto num : vec)
		cnt += (target / num);
	
	return cnt >= m;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n >> m;
	vector<long long> vec(n);
	for (int i = 0; i < n; i++)
		cin >> vec[i];
	
	sort(vec.begin(), vec.end());
	
	long long left = 0;
	long long right = vec[0] * m;

	while (left < right) {
		long long mid = (left + right) / 2;

		if (check_count(mid, vec)) {
			right = mid;
		} else {
			left = mid+1;
		}
	}

	cout << left;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
import java.io.*;

// 매개변수탐색 최소값
class Solution {

    static boolean check(long target, int n, int[] times) {
        long cnt = 0;
        for (int i = 0; i < times.length; i++) {
            cnt += (target / times[i]);
            
            if (cnt >= n) return true;
        }
        
        return false;
    }
    
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long answer = 0;
        long left = 0;
        long right = times[times.length-1]*(long)n;
        
        while (left <= right) {
            long mid = (left + right) / 2;
            
            if (check(mid, n, times)) {
                answer = mid;
                right = mid-1;
            } else {
                left = mid+1;
            }
        }
        return answer;
    }
}